### PR TITLE
Code Review and refactor.

### PR DESCRIPTION
**Pull request by JackWFinlay**
I saw your post on Reddit requesting a code review. I saw a few opportunities to refactor, and a few bugs to fix. Here's a rundown of my suggestions:

- Introduce `Constants.cs` for repeatedly used strings.
- Rename files and namespaces for consistency.
- Extract logic to `GildedRose.cs`.
- Fix logic error in `UpdateQuality()`: use `continue` instead of `return`. The latter causes any object in the list after 'Sulfuras...' to be skipped entirely.
- Remove `.vscode` directory. This shouldn't have been committed.
- Add `.vscode` to .gitignore. This prevents it being added again to git.
- Upgrade to dotnet core 3.1. Just became available, with LTS.
- Removed `GetItems()`, it was only used for testing and is pointless; you have a reference to the same set of items in the `items` List object at the beginning of each test anyway.
- Pull logic for each case into its own method.
- Add back in code in `Program.cs` to actually run the app. Enhanced to print details rather than just the object's type.

Regards,
Jack

### CODE REVIEW

**Comment by SSmale** [comment]
Thank you for the review, learnt a lot. 

With the GetItems() method, does this work because the Items list is passed by reference? if it was a primitive value would you need a similar method to get the change?

File: GildedRoseApp.Console/Program.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+                
+                foreach (Item item in items)
+                {
+                    System.Console.WriteLine($"{item.Name}, {item.SellIn}, {item.Quality}");
+                }
+                
+                System.Console.WriteLine();
+                
+                app.UpdateQuality();
+            }
```
Comment (SSmale) [inline GildedRoseApp.Console/Program.cs:41]: This is a cool addition to the original.

**Comment by JackWFinlay** [comment]
> With the GetItems() method, does this work because the Items list is passed by reference? if it was a primitive value would you need a similar method to get the change?

Yes, exactly. In the case of primitives, you would be best to return the values from the method, rather than exposing it publicly, though.

**Comment by SSmale** [comment]
> > With the GetItems() method, does this work because the Items list is passed by reference? if it was a primitive value would you need a similar method to get the change?
> 
> Yes, exactly. In the case of primitives, you would be best to return the values from the method, rather than exposing it publicly, though.

Would you be able to show me an example?

**Comment by JackWFinlay** [comment]
> > > With the GetItems() method, does this work because the Items list is passed by reference? if it was a primitive value would you need a similar method to get the change?
> > 
> > 
> > Yes, exactly. In the case of primitives, you would be best to return the values from the method, rather than exposing it publicly, though.
> 
> Would you be able to show me an example?

In the below, we are modifying the value that was passed, but because it's a value type, we return a value from the method. The original value isn't modified as it was passed by value.

```
public static void Main(string[] args)
{
    int someInt = 4;

    Console.WriteLine(someInt); // Outputs: 4

    int anotherInt = ChangeValueType(someInt);

    Console.WriteLine(someInt); // Outputs: 4
    Console.WriteLine(anotherInt); // Outputs: 5
}

private static int ChangeValueType(int value)
{
    value = 5;
    return value;
}
```

[See here for info on the ref keyword](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/ref), if you want to pass a value type by reference. I don't recommend it though.

[the out keyword does something similar](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/out-parameter-modifier)

File: GildedRoseApp.Console/Program.cs (hunk 1 above)
Comment (JackWFinlay) [reply GildedRoseApp.Console/Program.cs:41]: It's taken from the original Kata, modified to be more modern C#, [original is here](https://github.com/emilybache/GildedRose-Refactoring-Kata/blob/master/csharpcore/Program.cs#L43).

Comment (SSmale) [reply GildedRoseApp.Console/Program.cs:41]: ahh cool, this one was from the NotMyself version originally

**Comment by SSmale** [comment]
> > > > With the GetItems() method, does this work because the Items list is passed by reference? if it was a primitive value would you need a similar method to get the change?
> > > 
> > > 
> > > Yes, exactly. In the case of primitives, you would be best to return the values from the method, rather than exposing it publicly, though.
> > 
> > 
> > Would you be able to show me an example?
> 
> In the below, we are modifying the value that was passed, but because it's a value type, we return a value from the method. The original value isn't modified as it was passed by value.
> 
> ```
> public static void Main(string[] args)
> {
>     int someInt = 4;
> 
>     Console.WriteLine(someInt); // Outputs: 4
> 
>     int anotherInt = ChangeValueType(someInt);
> 
>     Console.WriteLine(someInt); // Outputs: 4
>     Console.WriteLine(anotherInt); // Outputs: 5
> }
> 
> private static int ChangeValueType(int value)
> {
>     value = 5;
>     return value;
> }
> ```
> 
> [See here for info on the ref keyword](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/ref), if you want to pass a value type by reference. I don't recommend it though.
> 
> [the out keyword does something similar](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/out-parameter-modifier)

Sorry to ask so many questions, But this has been a great help. in the case of my AddItem(), how would you get the value if the ctor took an int instead?

**Comment by JackWFinlay** [comment]
> Sorry to ask so many questions, But this has been a great help. in the case of my AddItem(), how would you get the value if the ctor took an int instead?

In this case you could expose it as a public property. You could have done the same with the `Items` field in `GildedRose.cs`, if you wanted to. In my pull-request I am just highlighting that it could in fact be used without creating another reference.

The below illustrates an [auto-implemented propery](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/auto-implemented-properties). Note that in C# we don't genereally use getters and setters like you would see in Java e.g. `GetItems()`, or `SetItems()`. By providing the `{ get; }` you allow it to implement a getter for you. If you add `set;`, it would also allow you to set the value from elsewhere.

``` CSharp
public int MyInt { get; set; }
```
You can then use this with:
```CSharp
int foo = app.MyInt;
```
or 
```CSharp
app.MyInt = 5;
```

So, in following this, you could actually change 
``` CSharp
private readonly IList<Item> _items;
``` 
to 
``` CSharp
public IList<Item> Items { get; set; }
``` 
if you wanted to. I would recommend still passing the `Items` via the constructor, but if you do want the items to be accessible, you can use the above as well.

Hope this helps.

**Comment by SSmale** [comment]
Thank you so much for your feedback and answers.

**Pull request merged**